# Exercise

1.1 Create a 1-D array [0..9].

In [1]:
import numpy as np
from numpy.random import default_rng

In [26]:
def print_array_info(array):
    print("Array Information:")
    print(f"Data Type: {array.dtype}")
    print(f"Shape: {array.shape}")
    print(f"Number of Dimensions: {array.ndim}")
    print(f"Item Size (in bytes): {array.itemsize}")
    print(f"Number of Elements: {array.size}")
    print(f"Strides: {array.strides}")
    print(f"Flags: {array.flags}")
    print(f"Total Size (in bytes): {array.nbytes}")

In [3]:
a = np.arange(10)
print(a)

[0 1 2 3 4 5 6 7 8 9]


1.2 Create a 3×4 array with values 0..11.

In [9]:
a = np.arange(12).reshape((3, 4))
print(a)

[[ 0  1  2  3]
 [ 4  5  6  7]
 [ 8  9 10 11]]


1.3 Create a float array from list [1,2,3].

In [11]:
a = np.array([1,2,3], dtype=float)
print(a)

[1. 2. 3.]


1.4 Create zeros of shape (2,3).

In [14]:
a = np.zeros((2,3))
print(a)

[[0. 0. 0.]
 [0. 0. 0.]]


1.5 Create ones of shape (3,).

In [17]:
a = np.ones((3))
print(a)

[1. 1. 1.]


1.6 Create an empty array (uninitialized) shape (4,4).

In [18]:
a = np.empty((4,4))
print(a)

[[0.00000000e+000 0.00000000e+000 0.00000000e+000 0.00000000e+000]
 [0.00000000e+000 0.00000000e+000 0.00000000e+000 0.00000000e+000]
 [0.00000000e+000 0.00000000e+000 0.00000000e+000 0.00000000e+000]
 [0.00000000e+000 0.00000000e+000 0.00000000e+000 2.12199579e-314]]


1.7 Create linspace from 0 to 1 with 5 points.

In [20]:
a = np.linspace(0, 1, 5)
print(a)

[0.   0.25 0.5  0.75 1.  ]


1.8 Create integer array [10,20,30] using array.

In [22]:
a = np.array([10, 20, 30])
print(a)

[10 20 30]


1.9 Get shape, ndim, size, dtype of A.

In [27]:
print_array_info(a)

Array Information:
Data Type: int64
Shape: (3,)
Number of Dimensions: 1
Item Size (in bytes): 8
Number of Elements: 3
Strides: (8,)
Flags:   C_CONTIGUOUS : True
  F_CONTIGUOUS : True
  OWNDATA : True
  WRITEABLE : True
  ALIGNED : True
  WRITEBACKIFCOPY : False

Total Size (in bytes): 24


1.10 Index: get element 7 from 1-D a.

In [34]:
a = np.arange(1, 10)
print(a[6])

7


1.11 Index 2nd row, 3rd column of A.

In [72]:
a = np.arange(12).reshape((3, 4))
print(a)
print(a[1:2])

[[ 0  1  2  3]
 [ 4  5  6  7]
 [ 8  9 10 11]]
[[4 5 6 7]]


1.12 Slice: first two rows of A.

In [45]:
print(a[:-1])
print(a[:2, :])

[[0 1 2 3]
 [4 5 6 7]]
[[0 1 2 3]
 [4 5 6 7]]


1.13 Reverse columns of A.

In [52]:
print(a[::-1, ::-1]) #reversed columns and rows

[[11 10  9  8]
 [ 7  6  5  4]
 [ 3  2  1  0]]


1.14 Create boolean mask where a is even.

In [54]:
mask = a % 2 == 0
print(mask)

[[ True False  True False]
 [ True False  True False]
 [ True False  True False]]


1.15 Fancy indexing: select rows 0 and 2 of A.

In [55]:
print(a[[0,2]])

[[ 0  1  2  3]
 [ 8  9 10 11]]


1.16 Boolean indexing: elements >5 in A.

In [56]:
mask = a > 5
print(mask)

[[False False False False]
 [False False  True  True]
 [ True  True  True  True]]


1.17 Demonstrate view vs copy: slice vs copy().

Note: modifying v may change A; modifying copy does not.

In [57]:
v = a  # view in many cases
v[0,0] = 999
print(a)

[[999   1   2   3]
 [  4   5   6   7]
 [  8   9  10  11]]


In [60]:
copy = a.copy()
copy[0,0] = -1
print(a)

[[ 0  1  2  3]
 [ 4  5  6  7]
 [ 8  9 10 11]]


1.18 Get strides of A.

In [61]:
a.strides

(32, 8)

In [62]:
print(a.astype(np.float32))

[[ 0.  1.  2.  3.]
 [ 4.  5.  6.  7.]
 [ 8.  9. 10. 11.]]


1.20 Make an array of dtype bool from [True, False].

In [64]:
b = np.array([True, False])
print(b)
print(b.astype(np.int32))

[ True False]
[1 0]


1.21 Create eye(4) identity matrix.

In [66]:
a = np.eye(4)
print(a)

[[1. 0. 0. 0.]
 [0. 1. 0. 0.]
 [0. 0. 1. 0.]
 [0. 0. 0. 1.]]


1.22 Create diag([1,2,3]).

In [67]:
a = np.diag([1,2,3])
print(a)

[[1 0 0]
 [0 2 0]
 [0 0 3]]


1.23 Create vander([1,2,3], N=3).

In [68]:
a = np.vander([1, 2, 3], 3)
print(a)

[[1 1 1]
 [4 2 1]
 [9 3 1]]


1.24 Use fromiter to build array from generator i*i for i in 0..9.

In [69]:
a = np.fromiter((i*i for i in range(10)), dtype=int, count=10)
print(a)

[ 0  1  4  9 16 25 36 49 64 81]


1.25 Create array with fromfunction f(i,j)=i+j shape (3,3).

In [71]:
a = np.fromfunction(lambda i,j: i+j, (3,3), dtype=int)
print(a)

[[0 1 2]
 [1 2 3]
 [2 3 4]]


1.26 Flatten with ravel vs flatten.

In [73]:
print(a.ravel)
print(a.flatten)

<built-in method ravel of numpy.ndarray object at 0x00000283A6EAD6B0>
<built-in method flatten of numpy.ndarray object at 0x00000283A6EAD6B0>


1.27 Transpose A.

In [74]:
print(a.T)

[[ 0  4  8]
 [ 1  5  9]
 [ 2  6 10]
 [ 3  7 11]]


1.28 Create structured dtype (record) with fields ('x','y') ints.

In [80]:
dt = np.dtype([('x', np.int32), ('y', np.int32)])
rec = np.array([(1,2),(3,4)], dtype=dt)
print(rec)
rec.reshape(2, 1, 1)

[(1, 2) (3, 4)]


array([[[(1, 2)]],

       [[(3, 4)]]], dtype=[('x', '<i4'), ('y', '<i4')])

1.29 Create np.arange(12).reshape(3,4) and broadcast-add [0,1,2,3].

In [99]:
a = np.arange(12).reshape(3,4)
b = np.array([0, 1, 2, 3])
print(a+b)

[[ 0  2  4  6]
 [ 4  6  8 10]
 [ 8 10 12 14]]


1.30 Use np.broadcast_to to repeat [1,2] to shape (3,2).

In [90]:
b = np.broadcast_to(np.array([1,2,3]), (7,3))
print(b)
b.shape

[[1 2 3]
 [1 2 3]
 [1 2 3]
 [1 2 3]
 [1 2 3]
 [1 2 3]
 [1 2 3]]


(7, 3)

1.31 Use np.tile([1,2], (2,3)) & np.repeat

In [98]:
a = np.tile([1,2], (2,3))
print(a)
(a.shape, a.strides)

[[1 2 1 2 1 2]
 [1 2 1 2 1 2]]


((2, 6), (48, 8))

In [ ]:
c = np.repeat(a, [2])
print(c)
(c.shape, c.strides)

[[ 0  1  2  3  0  1  2  3]
 [ 4  5  6  7  4  5  6  7]
 [ 8  9 10 11  8  9 10 11]]


((3, 8),
 (64, 8),
   C_CONTIGUOUS : True
   F_CONTIGUOUS : False
   OWNDATA : False
   WRITEABLE : True
   ALIGNED : True
   WRITEBACKIFCOPY : False)

1.32 Create 3×3 random ints 0..9 using rng.

In [123]:
rng = default_rng(44)
c = rng.integers(0, 10, size = (3, 3))
d = np.random.randint(0, 10, size = (3, 3))
print(c)
print(d)

[[6 1 8]
 [2 6 4]
 [0 9 5]]
[[3 8 7]
 [4 8 8]
 [3 7 6]]


1.33 Create np.arange(9).reshape(3,3) and compute a.flat[4].

In [135]:
b = np.arange(9).reshape(3,3)
print(b)
print(b.flat[1].shape)

[[0 1 2]
 [3 4 5]
 [6 7 8]]
()


1.34 Create np.mgrid[0:3, 0:4].

In [180]:
b = np.mgrid[0:20:3, 0:20:3]
print(b)

[[[ 0  0  0  0  0  0  0]
  [ 3  3  3  3  3  3  3]
  [ 6  6  6  6  6  6  6]
  [ 9  9  9  9  9  9  9]
  [12 12 12 12 12 12 12]
  [15 15 15 15 15 15 15]
  [18 18 18 18 18 18 18]]

 [[ 0  3  6  9 12 15 18]
  [ 0  3  6  9 12 15 18]
  [ 0  3  6  9 12 15 18]
  [ 0  3  6  9 12 15 18]
  [ 0  3  6  9 12 15 18]
  [ 0  3  6  9 12 15 18]
  [ 0  3  6  9 12 15 18]]]


some np.meshgrid

In [181]:
x = np.array([1, 2, 3])
y = np.array([10, 20])

print(np.meshgrid(x, y))

(array([[1, 2, 3],
       [1, 2, 3]]), array([[10, 10, 10],
       [20, 20, 20]]))


In [186]:
x = np.linspace(-2, 2, 5)
y = np.linspace(-3, 3, 7)

X, Y = np.meshgrid(x, y)
print(X.T.shape)

print(Y.shape)

Z = X + Y

print(Z.shape)

(5, 7)
(7, 5)
(7, 5)


1.35 Use np.ogrid to create (3,1) and (1,4) open grids.

In [185]:
x, y = np.ogrid[0:3, 0:4]
(x.shape, x.ndim , y.shape, y.ndim)

((3, 1), 2, (1, 4), 2)

1.36 Use np.tril on a 4×4 range matrix.

In [192]:
M = np.arange(16).reshape(4,4)
print(np.tril(M))
print(np.triu(M))

[[ 0  0  0  0]
 [ 4  5  0  0]
 [ 8  9 10  0]
 [12 13 14 15]]
[[ 0  1  2  3]
 [ 0  5  6  7]
 [ 0  0 10 11]
 [ 0  0  0 15]]


1.37 Create zeros_like(A) and ones_like(A).

In [194]:
print(np.zeros_like(a))
print(np.ones_like(a))

[[0 0 0 0]
 [0 0 0 0]
 [0 0 0 0]]
[[1 1 1 1]
 [1 1 1 1]
 [1 1 1 1]]


1.38 Convert a 0-D scalar array np.array(5) to Python int.

In [205]:
c = np.array(5)
int(c)
print(c)


5


1.39 Memory view: show base attribute of a slice.

In [208]:
s = a[0:2]
s.base is a

False

1.40 Create astype with casting safe (default) vs forcing with astype(..., copy=False).

In [209]:
a.astype(np.float64, copy=False)

array([[ 0.,  1.,  2.,  3.],
       [ 4.,  5.,  6.,  7.],
       [ 8.,  9., 10., 11.]])

### Math Ops & reductions

2.1 Sum all elements of A.

In [211]:
print(np.sum(a))

66


2.2 Sum along axis.

In [212]:
print(np.sum(a, axis = 0))
print(np.sum(a, axis = 1))

[12 15 18 21]
[ 6 22 38]


2.3 Sum with dtype int64 to avoid overflow.

In [216]:
print(np.sum(np.int64(a)))
print(np.sum(a.astype(np.int32)))
np.sum(a.astype(np.int32), dtype=np.int64)

66
66


np.int64(66)

2.4 Mean of A, axis 1, keepdims True.

In [232]:
print(np.mean(a, axis=1, keepdims=True))
print(np.mean(np.array([1, np.nan, 5])))
print(np.nanmean(np.array([1, np.nan, 5])))

[[1.5]
 [5.5]
 [9.5]]
nan
3.0


2.5 Compute np.min and np.max of A.

In [218]:
print(np.min(a))
print(np.max(a))

0
11


2.6 Compute absolute values of np.array([-1,-2,3]).

In [219]:
print(np.abs(np.array([-1,-2,3])))

[1 2 3]


2.7 Use np.exp on np.array([0,1,2]).

In [221]:
print(np.exp(np.array([0, 1, 2])))

[1.         2.71828183 7.3890561 ]


2.8 Compute sin and cos of [0, pi/2, pi].

In [223]:
print(np.round(np.sin(np.array([0, np.pi/2, np.pi]))))
print(np.round(np.cos(np.array([0, np.pi/2, np.pi]))))

[0. 1. 0.]
[ 1.  0. -1.]


2.9 np.allclose compare [0.1+0.2] vs [0.3] with default tolerances.

In [228]:
np.allclose(np.array([1 - 0.7]), np.array([0.3]))

True

2.10 Use np.nansum to sum [1, np.nan, 3].

In [231]:
print(np.nansum(np.array([1, np.nan, 3])))
print(np.sum(np.array([1, np.nan, 10])))

4.0
nan


2.10 Use np.nansum to sum [1, np.nan, 3].

In [234]:
np.nansum(np.array([1, np.nan, 3]))


np.float64(4.0)

2.12 Compute L2 norm of vector [3,4] using np.sqrt(np.sum(x**2)).

In [236]:
print(np.sqrt(np.sum(a**2)))

22.494443758403985


2.13 Compute elementwise exp(-x) for x = np.linspace(0,5,6).

In [248]:
b = np.linspace(0, 5, 6)
print(np.exp(-b))

[1.         0.36787944 0.13533528 0.04978707 0.01831564 0.00673795]


2.14 Compute softmax stable: implement numerically stable softmax for logits.

In [ ]:
# simple softmax
logits = np.array([0.0000000000000000000000000000000000000001, 1.0, 0.1])
exp_logits = np.exp(logits)
softmax = exp_logits / np.sum(exp_logits)
print(softmax)

[0.20732037 0.56355519 0.22912444]


In [ ]:
# stable softmax
logits = np.array([0.0000000000000000000000000000000000000001, 1.0, 0.1])
m = np.max(logits)
exp_logits = np.exp(logits-m)
softmax = exp_logits / exp_logits.sum()
print(softmax)

[0.20732037 0.56355519 0.22912444]


2.15 Compute np.sum in chunks to reduce error (Kahan not in numpy): show difference by using float64.

In [252]:
b = (1e8 * np.ones(1000)) + np.linspace(0,999,1000)
using_64 = np.sum(b, dtype=np.float64)
using_32 = np.sum(b.astype(np.float32), dtype=np.float32)
print(using_64)
print(using_32)

100000499500.0
1.00000506e+11


## Questions to solve

1. Create a NumPy array from the list [2, 4, 6, 8].

In [3]:
a = np.array([2, 4, 6, 8])
print(a)

[2 4 6 8]


2. Create an array from 0 to 9 using np.arange().

In [4]:
a = np.arange(10)
print(a)

[0 1 2 3 4 5 6 7 8 9]


3. Create a 3×3 array filled with zeros.

In [7]:
a = np.zeros((3, 3))
print(a)

[[0. 0. 0.]
 [0. 0. 0.]
 [0. 0. 0.]]


4. Create a 2×4 array filled with ones.

In [ ]:
a =np.ones((2, 4))
print()

5. Generate 7 evenly spaced values between 5 and 10.

In [10]:
a =  np.linspace(1,10, 3)
print(a)

[ 1.   5.5 10. ]


6. Reshape np.arange(12) into a 3×4 array.

In [21]:
a = np.arange(12).reshape(3,4)
print(a)

[[ 0  1  2  3]
 [ 4  5  6  7]
 [ 8  9 10 11]]


7. Flatten a 2D array a.

In [14]:
b = a.flatten()
print(b)

[ 0  1  2  3  4  5  6  7  8  9 10 11]


8. Ravel a 2D array a.

In [16]:
print(a.ravel())

[ 0  1  2  3  4  5  6  7  8  9 10 11]


9. Get the sum of all elements in array a.

In [17]:
print(np.sum(a))

66


10. Get the mean of a 1D array arr.

In [20]:
print(np.mean(a))

5.5


11. Apply the sine function on array x.


In [22]:
print(np.sin(a))

[[ 0.          0.84147098  0.90929743  0.14112001]
 [-0.7568025  -0.95892427 -0.2794155   0.6569866 ]
 [ 0.98935825  0.41211849 -0.54402111 -0.99999021]]


12. Get the dot product of arrays a and b.

In [33]:
b = np.arange(4)
print(b)
print(a)
c = np.dot(a, b)
print(c)

[0 1 2 3]
[[ 0  1  2  3]
 [ 4  5  6  7]
 [ 8  9 10 11]]
[14 38 62]


13. Generate a 3×3 array of random integers from 1 to 20.

In [38]:
c = np.random.randint(1, 20, (3,3,3))
print(c)

[[[ 2  8  5]
  [16  6 15]
  [19 19 17]]

 [[17 12 19]
  [ 8  8 14]
  [ 4  2 12]]

 [[12  7 14]
  [ 2 11 18]
  [13  7  3]]]


14. Access the first element of a 3D array x using [0][0][0].

In [47]:
print(c[2][0][0])

12


15. Slice the entire array using [:, :, :].


Purpose	Example

single element	A[i,j,k]

whole layer	A[i,:,:]

row across layers	A[:,j,:]

col across layers	A[:,:,k]

subcube	A[1:3, 0:2, 0:2]

reverse	A[::-1, :, :]

step slices	A[::2, ::2, ::2]

fancy indexing	A[[0,2], :, :]

In [56]:
print(c[2::, 1:2, 2:5])

[[[18]]]
